src/contracts/GPv2UniswapRouter.sol

```diff
@@ -29,6 +29,62 @@ contract GPv2UniswapRouter is UniswapV2Library {
         factory = factory_;
     }
 
+    /// @dev Execute a GPv2 settlement with a single trade against a Uniswap
+    /// path.
+    ///
+    /// This method performs no input validation, instead it performs on-chain
+    /// computation of the swap amounts in order to encode interactions and set
+    /// clearing prices such as the trader receives positive slippage. Verifying
+    /// the signature and validity of the trade (limit prices, expiry, etc.) is
+    /// done by the settlement contract.
+    ///
+    /// @param path The Uniswap route for trading.
+    /// @param trade The GPv2 trade to execute.
+    function settleSwap(IERC20[] calldata path, GPv2Trade.Data calldata trade)
+        external
+    {
+        uint256 tokenCount = path.length;
+        require(tokenCount > 1, "GPv2: invalid path");
+        require(
+            trade.sellTokenIndex == 0 && trade.buyTokenIndex == tokenCount - 1,
+            "GPv2: invalid trade for path"
+        );
+
+        uint256[] memory amounts;
+        {
+            (bytes32 kind, , ) = trade.flags.extractFlags();
+            amounts = kind == GPv2Order.SELL
+                ? getAmountsOut(factory, trade.sellAmount, path)
+                : getAmountsIn(factory, trade.buyAmount, path);
+        }
+
+        uint256[] memory prices = new uint256[](tokenCount);
+        prices[0] = amounts[tokenCount - 1];
+        prices[tokenCount - 1] = amounts[0];
```

Comment:
Maybe we can also validate the user's limit price here so we don't go through the trouble of sending funds back and forth before realising the limit price is not satisfied?

Reply:
In the settlement, after recovering the signature, the first thing that happens is limit price validation. It is done while the transfer amounts are being computed, but before they are executed.

This is a nice suggestion. I think to decide if we actually want this we need to:
- compute the amount of additional gas for the check
- compute the additional cost of reverting because of slippage
- estimate a failure rate
- compare `check_gas_cost * success_rate` with `extra_revert_cost * failure_rate`

Reply:
Given we fail "early enough" in the settlement function that's fine and I think the extra logic is not worth it.

Reply:
This will be enforced by the suggested limit amount :point_up: